feat(empty): Support better customization with semantic classNames/styles as function

Susuperli opened the pull request:
[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE_CN.md?plain=1)

### 🤔 This is a ...

- [x] 🆕 New feature
- [ ] 🐞 Bug fix
- [ ] 📝 Site / documentation improvement
- [ ] 📽️ Demo improvement
- [ ] 💄 Component style improvement
- [ ] 🤖 TypeScript definition improvement
- [ ] 📦 Bundle size optimization
- [ ] ⚡️ Performance optimization
- [ ] ⭐️ Feature enhancement
- [ ] 🌐 Internationalization
- [ ] 🛠 Refactoring
- [ ] 🎨 Code style optimization
- [ ] ✅ Test Case
- [ ] 🔀 Branch merge
- [ ] ⏩ Workflow
- [ ] ⌨️ Accessibility improvement
- [ ] ❓ Other (about what?)

### 🔗 Related Issues

> - https://github.com/ant-design/ant-design/discussions/54505

### 💡 Background and Solution

> - The specific problem to be addressed.
> - List the final API implementation and usage if needed.
> - If there are UI/interaction changes, consider providing screenshots or GIFs.

### 📝 Change Log

> - Read [Keep a Changelog](https://keepachangelog.com/en/1.1.0/)! Track your changes, like a cat tracks a laser pointer.
> - Describe the impact of the changes on developers, not the solution approach.
> - Reference: https://ant.design/changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |Support better customization with semantic classNames/styles as function|
| 🇨🇳 Chinese |使用语义类名/样式作为函数，支持更好的自定义|

Susuperli committed:
feat(empty): add semantic customization support based on next branch

- Add classNames and styles props for semantic customization
- Support four semantic parts: root, image, description, footer
- Use advanced useMergeSemantic hook with type parameters
- Add SemanticClassNamesType and SemanticStylesType utility types
- Implement mergedProps pattern for conditional semantic logic
- Add comprehensive semantic demo with style-class example
- Add ConfigProvider-level test for Empty semantic customization
- Update API documentation in both English and Chinese
- Follow PR #54917 pattern for consistent semantic implementation
- Based on next branch with latest semantic infrastructure

Susuperli committed:
test: add snapshots for Empty semantic style-class demo

- Add snapshot for style-class.tsx demo in demo.test.ts.snap
- Add snapshot for style-class.tsx demo in demo-extend.test.ts.snap
- Snapshots include semantic classNames and styles verification:
  * custom-empty-root with backgroundColor, padding, borderRadius
  * custom-empty-image with filter grayscale
  * custom-empty-description with color and fontWeight
  * custom-empty-footer with marginTop
- Verify semantic customization is properly applied to DOM elements
- Complete test coverage for Empty component semantic features

Susuperli committed:
fix: improve Empty semantic implementation based on review feedback

- Add default semantic classNames for image, description, footer parts
- Update API documentation with detailed descriptions and function type support
- Follow FloatButton pattern for consistent semantic implementation
- Improve documentation clarity for classNames and styles props
- Support both object and function-based semantic customization
- Ensure proper semantic class application to DOM elements

Susuperli committed:
fix: address PR review feedback for Empty semantic implementation

- Remove ineffective React.useMemo for mergedProps, use simple object spread
- Add comprehensive function-based semantic tests in describe('semantic classNames/styles')
- Add ConfigProvider-level function-based semantic tests
- Enhance demo to showcase both object and function-based semantic customization
- Add interactive demo with Switch to demonstrate dynamic function-based styling
- Test function parameter passing and prop dependency handling
- Ensure complete test coverage for both object and function semantic patterns

Addresses gemini-code-assist bot feedback:
1. Fixed ineffective useMemo with props dependency
2. Added function-based semantic test coverage
3. Enhanced demo to show function-based capabilities

Susuperli committed:
refactor: align Empty semantic tests with FloatButton PR #54917 pattern

- Simplify semantic test cases to match FloatButton implementation
- Use more focused test structure with rerender pattern
- Add proper EmptyProps type import for type safety
- Follow exact naming convention from FloatButton tests:
  * 'should apply dynamic classNames and styles from props function'
  * 'should apply object classNames and styles'
- Remove overly complex test cases, keep essential functionality tests
- Ensure test structure matches established Ant Design patterns

Based on FloatButton semantic tests analysis from PR #54917

Susuperli committed:
docs: improve Empty semantic API documentation and remove ConfigProvider tests

- Update Chinese documentation for classNames and styles properties
- Use more concise and consistent descriptions matching Ant Design standards
- Update version from 5.23.0 to 6.0.0 for semantic API
- Add missing documentation for style-class demo in both languages
- Remove ConfigProvider-level Empty semantic tests as requested
- Keep semantic tests focused in Empty component test file only

Changes:
- components/empty/index.zh-CN.md: Improved API descriptions
- components/empty/demo/style-class.md: Added demo documentation
- components/config-provider/__tests__/style.test.tsx: Removed Empty tests

Susuperli committed:
refactor: clean up Empty semantic tests and demo

- Remove remaining Empty semantic tests from ConfigProvider test file
- Simplify style-class demo to basic implementation
- Remove demo documentation to keep it minimal
- Focus semantic tests in Empty component test file only

This ensures cleaner separation of concerns and avoids test duplication.

Susuperli committed:
fix: resolve Empty semantic classNames duplication and update documentation

Susuperli committed:
Merge branch 'next' into feat/empty-semantic-next-based

Meet-student commented on components/empty/demo/style-class.tsx:
```diff
@@ -0,0 +1,88 @@
+  return (
+    <div>
+      <div style={{ marginBottom: 16 }}>
+        <Switch
```
这个切换逻辑不需要， demo 里太复杂了

4个不同的写法 放在 demo 里就行了

Meet-student commented on components/empty/demo/style-class.tsx:
```diff
@@ -0,0 +1,88 @@
+          description: 'custom-empty-description',
+          footer: 'custom-empty-footer',
+        }}
+        styles={{
```
这里 包裹下面的， 都写在上面封装对象和函数传入。

Susuperli commented:
> 改下comment 内容 @Susuperli

好的，刚才点错了😂

Susuperli committed:
refactor: address Meet-student review feedback for Empty semantic demo

- Fix documentation: change 'buttons' to 'empty' in both Chinese and English
- Remove iframe attribute from demo documentation
- Simplify demo to show 4 different usage patterns without complex toggle logic
- Add comprehensive examples: object/function forms with external/inline definitions
- Fix TypeScript warnings and improve code clarity

Susuperli commented:
> 改下comment 内容 @Susuperli

done

Meet-student commented on components/empty/demo/style-class.tsx:
```diff
@@ -0,0 +1,110 @@
+    <h4 style={{ marginTop: 32 }}>3. Inline object classNames and styles:</h4>
+    <Empty
+      image={Empty.PRESENTED_IMAGE_SIMPLE}
+      classNames={{
```
这里也写上面

Meet-student commented on components/empty/demo/style-class.tsx:
```diff
@@ -0,0 +1,110 @@
+        root: info.props.description ? 'has-description' : 'no-description',
+        description: 'custom-desc',
+      })}
+      styles={(info) => ({
```
这里也写上面

Susuperli committed:
refactor: address Meet-student review feedback and optimize demo structure

- Remove version numbers from API documentation (classNames/styles)
- Simplify demo to show 4 separate examples for each semantic property
- Add comprehensive ConfigProvider tests for function-based semantic usage
- Optimize demo structure: separate objectClassNames, objectStyles, functionClassNames, functionStyles
- Each demo example focuses on single property for better clarity

Susuperli committed:
Merge branch 'next' into feat/empty-semantic-next-based

Susuperli committed:
Merge branch 'next' into feat/empty-semantic-next-based

Susuperli committed:
refactor(empty): Optimize the type definitions and style types in the Empty component demo.

thinkasany commented on components/empty/index.tsx:
```diff
@@ -110,25 +134,16 @@ const Empty: CompoundedComponent = (props) => {
-          className={classNames(`${prefixCls}-footer`, mergedClassNames.footer)}
-          style={mergedStyles.footer}
-        >
+        <div className={mergedClassNames.footer} style={mergedStyles.footer}>
```
useMergeSemantic 里面没有合并 ``` `${prefixCls}-dom```,  也有用户会喜欢直接用 css 搭配全局自定义prefix 去魔改样式的，``` .xxx-footer```

thinkasany replied:
而且 css-in-js 不是要用到这边么

Susuperli replied:
我又研究了一下 改了改

Susuperli committed:
fix: ensure CSS-in-JS prefix classes are applied to Empty semantic elements

- Add manual prefix class application for image, description, and footer elements
- Remove emptyClassNames from useMergeSemantic to avoid duplication
- Ensure CSS-in-JS styles can properly target semantic elements
- Fix issue where custom prefixCls wouldn't work with CSS select

Susuperli committed:
test: Update snapshot

Susuperli committed:
Merge branch 'next' into feat/empty-semantic-next-based

Susuperli committed:
refactor(components/empty): Remove unused merge property code

Susuperli committed:
Merge branch 'next' into feat/empty-semantic-next-based

Susuperli committed:
Merge branch 'next' into feat/empty-semantic-next-based

li-jia-nan committed:
Merge branch 'next' into feat/empty-semantic-next-based

Susuperli committed:
Merge branch 'next' into feat/empty-semantic-next-based

Susuperli committed:
Merge branch 'next' into feat/empty-semantic-next-based

thinkasany commented:
@Susuperli  可以不用点击同步，一般我们会帮你同步最新的next分支，由于是第一次 PR，PR 的 ci 会被阻止，需要人工 ap 才运行。

主要是需要等待核心成员验收，我们看了大概没问题了（ci完整通过的话）。就只需要耐心等待即可。

也是希望到时候豆酱老师有空来看的时候，ci 是一个完整的通过状态，就可以直接合了，不然还得等待，时间确实有久。

感谢你的贡献～

Susuperli commented:
okk



---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 日期 | 2025年09月19日 19:49 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.***>***@***.***> |
| 主题 | Re: [ant-design/ant-design] feat(empty): Support better customization with semantic classNames/styles as function (PR #55007) |
thinkasany left a comment (ant-design/ant-design#55007)

@Susuperli 可以不用点击同步，一般我们会帮你同步最新的next分支，由于是第一次 PR，PR 的 ci 会被阻止，需要人工 ap 才运行。

主要是需要等待核心成员验收，我们看了大概没问题了（ci完整通过的话）。就只需要耐心等待即可。

也是希望到时候豆酱老师有空来看的时候，ci 是一个完整的通过状态，就可以直接合了，不然还得等待，时间确实有久。

感谢你的贡献～

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Meet-student commented on components/empty/demo/style-class.md:
```diff
@@ -0,0 +1,7 @@
+## zh-CN
+
+通过 `classNames` 和 `styles` 传入对象/函数可以自定义空状态的[语义化结构](#semantic-dom)样式。
```
```suggestion
通过 `classNames` 和 `styles` 传入对象/函数可以自定义 Empty 的[语义化结构](#semantic-dom)样式。
```

meet-student committed:
Update components/empty/demo/style-class.md

Signed-off-by: 遇见同学 <1875694521@qq.com>

meet-student committed:
Update components/empty/demo/style-class.md

Signed-off-by: 遇见同学 <1875694521@qq.com>

thinkasany commented:
@Meet-student 更新完记得 ap ci，第一次 PR 的同学，ci不会自动运行。

Susuperli committed:
Merge branch 'next' into feat/empty-semantic-next-based

Susuperli committed:
refactor(empty): simplify style-class demo following pagination pattern

- Keep four essential objects: classNamesObject, classNamesFn, stylesObject, stylesFn
- Use same classNames across examples with sharedClassNames
- Differentiate styles based on props in stylesFn
- Reduce code from 115 to 68 lines (40% reduction)
- Add clear English comments for each object
- Update snapshots for new structure
- Follow pagination demo concise style

Susuperli commented on components/empty/demo/style-class.tsx:
```diff
@@ -0,0 +1,84 @@
+};
+
+// Object-based styles
+const objectStyles = {
```
改了一下看看

Susuperli committed:
test(empty): update snapshots after style-class demo refactor

- Update demo.test.ts.snap with new simplified structure
- Update demo-extend.test.ts.snap accordingly
- Snapshots now reflect the refactored demo with 4 examples
- All tests passing with updated snapshots

Susuperli committed:
Merge branch 'next' into feat/empty-semantic-next-based

Meet-student commented on components/empty/demo/style-class.tsx:
```diff
@@ -0,0 +1,84 @@
+import type { EmptyProps } from '..';
+import { createStyles } from 'antd-style';
+
+const useStyle = createStyles(({ css }) => ({
```
demo 参考下这个精简下。样式搭配和谐一些

https://github.com/ant-design/ant-design/pull/54957/files#diff-71c24f063ec1a30f721c1a2bfb648454753dfa9f3b907e9e87fe7302472b3727

classNames 使用同一个，然后 styles 根据 props (尽量选择 type、size、variant)做一下区分

Susuperli committed:
Merge branch 'next' into feat/empty-semantic-next-based

Susuperli committed:
refactor(empty): simplify style-class demo

- Remove classNamesObject and classNamesFn as suggested 'use style instead'
- Reduce to 2 examples as requested 'use two examples only'
- Keep Object styles and Function styles to demonstrate styles usage
- Update snapshots accordingly
